File: neetcode all and misc questions/Trees/count_good_nodes_in_binary_tree.py
"""
https://leetcode.com/problems/count-good-nodes-in-binary-tree/

Time: O(N)
Space: No additional Space (O(1))

Key Concept:
1) My mistake was thinking you need to track the entire path, but really you only need to
keep track of the current maximum in the recursive call stack so far. 

2) Whenever we see a node greater than our current max, we update the current max and
increment the final global result by one since this current node would be considered "good"

For example, in the following tree

        3
     1      4
  3       1   5

Left side:
3 is a good node, because it's current max (the root 3) is greater than or equal to 3, increment res to 1

3 -> 1, 1 is not a good node, because the node <= current max (3) 

1 -> 3, 3 is a good node, because the node >= current Max (3), increment res to 2

So far in this path, there's 2 nodes, but we only needed to track what the current max was through the recursion

Right side:
3 -> 4, 4 is a good node, because the node >= current Max (3), we also update the current max to 4, increment res to 3

4 -> 5, 5 is a good node, because the node >= current Max (5), we also update the current max to 5, increment res to 4

4 -> 1, 1 is not a good node, because the node < current Max (5)

Final result is 4

"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        """
        Revisited 10/1/2024, same solution as below
        at a given node, it would be helpful if we knew the values of the nodes
        before it on the path. Specifically, we're interested in what the "max"
        we've found so far on this path. Whenever the current max is updated, this means
        this is a "good" node, as there shouldn't be any node we've visited in this path
        with a value greater than this current max.
        
        Traverse the tree, passing down the current max with each recursive call. Whenever
        the current node val >= currentMax (meaning previous nodes CAN be equal to the current max but not exceeding it), increment the count of good nodes by 1,
        and in the next recursive call, pass in the
        current node val instead of currentMax to update it
        """
        self.count = 0
        def search(root, currentMax):
            if root:
                curMax = currentMax
                if root.val >= currentMax:
                    self.count+=1
                    curMax = root.val
                search(root.left, curMax)
                search(root.right, curMax)
        search(root, float("-inf"))
        return self.count            

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.res = 0
        if not root:
            return 0
        def helper(root, curMax):
            if root:
                if root.val >= curMax:
                    self.res += 1
                    curMax = max(curMax, root.val)
                helper(root.left, curMax)
                helper(root.right, curMax)
        helper(root, root.val)
        return self.res

File: node_js_ds_and_algos_implementations/trees/count_good_nodes_in_binary_tree.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root) {
        /*
        Revisited 5/7/2026
        storing the current max while we traverse the tree,
        if the node is greater or equal to the current max, that means
        that there is no node in the path that has a value greater than
        the current, so this would be considered a "good" node 

        for example
           1
         1   1

        technically, all 3 nodes here are good nodes, because they are all equal,
        but no value is greater than the other, which explains why its >= and not just >

        Time: O(N)
        Space: O(1)
        */
        if (!root){
            return 0
        }
        let res = 0
        const search = (node, curMax) => {
            if (node){
                let newMax = curMax
                if (node.val >= curMax){
                    res += 1
                    newMax = node.val
                }
                search(node.left, newMax)
                search(node.right, newMax)            
            }
        }
        search(root, root.val)
        return res
    }
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    if (!root){
        return []
    }
    let nodes = []
	// keep track of the current max as a function param 
	// while we traverse the tree.
    var search = function(root, curMax){
        if (root){
        	// if the current root's val is >= to the current max,
        	// this is a "good" node, so we append to our list,
            if (root.val >= curMax){
                nodes.push(root.val)
	        	// update the current max by passing in our current node val
                search(root.left, root.val)
                search(root.right, root.val)
            }
            else {
            	// if we didn't find a greater value, just pass in the current max
                search(root.left, curMax)
                search(root.right, curMax)
            }

        }
    }
    search(root, root.val)
    // take only the length of the list
    return nodes.length
};